File: e2e/node/tests/robots.test.ts
import { describe, it, expect } from "vitest";
import { scrape, createEngine } from "@kreuzberg/kreuzcrawl";

describe("robots", () => {
	it("robots_allow_all: Permissive robots.txt allows all paths", async () => {
		const engineConfig = { respect_robots_txt: true };
		const engine = createEngine(engineConfig);
		const url = `${process.env.MOCK_SERVER_URL}/fixtures/robots_allow_all`;
		const result = await scrape(engine, url);
		expect(result.isAllowed).toBe(true);
	});

	it("robots_allow_override: Allow directive overrides Disallow for specific paths", async () => {
		const engineConfig = { respect_robots_txt: true };
		const engine = createEngine(engineConfig);
		const url = `${process.env.MOCK_SERVER_URL}/fixtures/robots_allow_override`;
		const result = await scrape(engine, url);
		expect(result.isAllowed).toBe(true);
	});

	it("robots_comments_handling: Correctly parses robots.txt with inline and line comments", async () => {
		const engineConfig = { respect_robots_txt: true, user_agent: "kreuzcrawl" };
		const engine = createEngine(engineConfig);
		const url = `${process.env.MOCK_SERVER_URL}/fixtures/robots_comments_handling`;
		const result = await scrape(engine, url);
		expect(result.isAllowed).toBe(true);
	});

	it("robots_meta_nofollow: Detects nofollow meta robots tag and skips link extraction", async () => {
		const engineConfig = { respect_robots_txt: true };
		const engine = createEngine(engineConfig);
		const url = `${process.env.MOCK_SERVER_URL}/fixtures/robots_meta_nofollow`;
		const result = await scrape(engine, url);
		expect(result.nofollowDetected).toBe(true);
	});

	it("robots_meta_noindex: Detects noindex meta robots tag in HTML page", async () => {
		const engineConfig = { respect_robots_txt: true };
		const engine = createEngine(engineConfig);
		const url = `${process.env.MOCK_SERVER_URL}/fixtures/robots_meta_noindex`;
		const result = await scrape(engine, url);
		expect(result.noindexDetected).toBe(true);
	});

	it("robots_missing_404: Missing robots.txt (404) allows all crawling", async () => {
		const engineConfig = { respect_robots_txt: true };
		const engine = createEngine(engineConfig);
		const url = `${process.env.MOCK_SERVER_URL}/fixtures/robots_missing_404`;
		const result = await scrape(engine, url);
		expect(result.isAllowed).toBe(true);
	});

	it("robots_multiple_user_agents: Picks the most specific user-agent block from robots.txt", async () => {
		const engineConfig = { respect_robots_txt: true, user_agent: "SpecificBot" };
		const engine = createEngine(engineConfig);
		const url = `${process.env.MOCK_SERVER_URL}/fixtures/robots_multiple_user_agents`;
		const result = await scrape(engine, url);
		expect(result.isAllowed).toBe(true);
	});

	it("robots_sitemap_directive: Discovers sitemap URL from Sitemap directive in robots.txt", async () => {
		const engineConfig = { respect_robots_txt: true };
		const engine = createEngine(engineConfig);
		const url = `${process.env.MOCK_SERVER_URL}/fixtures/robots_sitemap_directive`;
		const result = await scrape(engine, url);
		expect(result.isAllowed).toBe(true);
	});

	it("robots_x_robots_tag: Respects X-Robots-Tag HTTP header directives", async () => {
		const engineConfig = { respect_robots_txt: true };
		const engine = createEngine(engineConfig);
		const url = `${process.env.MOCK_SERVER_URL}/fixtures/robots_x_robots_tag`;
		const result = await scrape(engine, url);
		expect(result.xRobotsTag.trim()).toBe("noindex, nofollow");
		expect(result.noindexDetected).toBe(true);
		expect(result.nofollowDetected).toBe(true);
	});
});

File: e2e/elixir/test/robots_test.exs
# E2e tests for category: robots
defmodule E2e.RobotsTest do
  use ExUnit.Case, async: true

  describe "robots_allow_all" do
    test "Permissive robots.txt allows all paths" do
      engine = Kreuzcrawl.create_engine!(nil)
      result = Kreuzcrawl.scrape!(engine, "")
      assert String.trim(result.is_allowed) == true
    end
  end

  describe "robots_allow_override" do
    test "Allow directive overrides Disallow for specific paths" do
      engine = Kreuzcrawl.create_engine!(nil)
      result = Kreuzcrawl.scrape!(engine, "")
      assert String.trim(result.is_allowed) == true
    end
  end

  describe "robots_comments_handling" do
    test "Correctly parses robots.txt with inline and line comments" do
      engine = Kreuzcrawl.create_engine!(nil)
      result = Kreuzcrawl.scrape!(engine, "")
      assert String.trim(result.is_allowed) == true
    end
  end

  describe "robots_crawl_delay" do
    test "Respects crawl-delay directive from robots.txt" do
      engine = Kreuzcrawl.create_engine!(nil)
      result = Kreuzcrawl.scrape!(engine, "")
      assert String.trim(result.crawl_delay) == 2
    end
  end

  describe "robots_disallow_path" do
    test "Robots.txt disallows specific paths" do
      engine = Kreuzcrawl.create_engine!(nil)
      result = Kreuzcrawl.scrape!(engine, "")
      assert String.trim(result.is_allowed) == false
    end
  end

  describe "robots_meta_nofollow" do
    test "Detects nofollow meta robots tag and skips link extraction" do
      engine = Kreuzcrawl.create_engine!(nil)
      result = Kreuzcrawl.scrape!(engine, "")
      assert String.trim(result.nofollow_detected) == true
    end
  end

  describe "robots_meta_noindex" do
    test "Detects noindex meta robots tag in HTML page" do
      engine = Kreuzcrawl.create_engine!(nil)
      result = Kreuzcrawl.scrape!(engine, "")
      assert String.trim(result.noindex_detected) == true
    end
  end

  describe "robots_missing_404" do
    test "Missing robots.txt (404) allows all crawling" do
      engine = Kreuzcrawl.create_engine!(nil)
      result = Kreuzcrawl.scrape!(engine, "")
      assert String.trim(result.is_allowed) == true
    end
  end

  describe "robots_multiple_user_agents" do
    test "Picks the most specific user-agent block from robots.txt" do
      engine = Kreuzcrawl.create_engine!(nil)
      result = Kreuzcrawl.scrape!(engine, "")
      assert String.trim(result.is_allowed) == true
    end
  end

  describe "robots_request_rate" do
    test "Parses request-rate directive from robots.txt" do
      engine = Kreuzcrawl.create_engine!(nil)
      result = Kreuzcrawl.scrape!(engine, "")
      assert String.trim(result.crawl_delay) == 5
      assert String.trim(result.is_allowed) == true
    end
  end

  describe "robots_sitemap_directive" do
    test "Discovers sitemap URL from Sitemap directive in robots.txt" do
      engine = Kreuzcrawl.create_engine!(nil)
      result = Kreuzcrawl.scrape!(engine, "")
      assert String.trim(result.is_allowed) == true
    end
  end

  describe "robots_user_agent_specific" do
    test "Matches user-agent specific rules in robots.txt" do
      engine = Kreuzcrawl.create_engine!(nil)
      result = Kreuzcrawl.scrape!(engine, "")
      assert String.trim(result.is_allowed) == false
    end
  end

  describe "robots_wildcard_paths" do
    test "Handles wildcard Disallow patterns in robots.txt" do
      engine = Kreuzcrawl.create_engine!(nil)
      result = Kreuzcrawl.scrape!(engine, "")
      assert String.trim(result.is_allowed) == false
    end
  end

  describe "robots_x_robots_tag" do
    test "Respects X-Robots-Tag HTTP header directives" do
      engine = Kreuzcrawl.create_engine!(nil)
      result = Kreuzcrawl.scrape!(engine, "")
      assert String.trim(result.x_robots_tag) == "noindex, nofollow"
      assert String.trim(result.noindex_detected) == true
      assert String.trim(result.nofollow_detected) == true
    end
  end
end
